### PR TITLE
Tr/crux separate online solver

### PR DESCRIPTION
**Pull request by travitch**
This set of commits adds options to crux to let the user choose different solvers for path satisfiability checking and proving goals.  Changes:

- Users can select a separate solver for path satisfiability checking (`--path-sat-solver`)
- Goals can be proved using solvers that do not have an online mode (e.g., boolector and dreal)
- The goal solver can be forced into offline mode even if it supports online solving (useful for eventually solving in parallel)
- The default behavior (if only `--solver` is specified) should be the same as before
- Adds support for the STP solver in both online and offline modes

### CODE REVIEW

File: crux/src/Crux.hs [hunk 1]
```diff
@@ -117,53 +128,93 @@ showVersion l = outputLn ("crux version: " ++ Crux.version ++ ", " ++
...
+            Just s -> symCfg sym cvc4Timeout (floor (s * 1000))
+            Nothing -> return ()
+          k floatRepr sym
+        CCS.Z3Online -> withZ3OnlineBackend floatRepr nonceGen ProduceUnsatCores $ \sym -> do
+          case goalTimeout cruxOpts of
+            Just s -> symCfg sym z3Timeout (floor (s * 1000))
+            Nothing -> return ()
+          k floatRepr sym
+        CCS.STPOnline ->
+          -- We don't have a timeout option for STP
```
Comment (robdockins) [inline crux/src/Crux.hs]: Issue a warning, perhaps, if `goalTimeout` is configured but we can't accomidate?

File: crux/src/Crux/Goal.hs [hunk 2]
```diff
@@ -129,24 +135,121 @@ proveToGoal _ allAsmps p pr =
...
+                   . (?outputConfig :: OutputConfig, sym ~ ExprBuilder t st fs)
+                  => WS.SolverAdapter st
+                  -> CruxOptions
+                  -> SimCtxt sym p
+                  -> Maybe (Goals (LPred sym asmp) (LPred sym ast))
+                  -> IO (Maybe (Goals (LPred sym asmp) (LPred sym ast, ProofResult (Either (LPred sym asmp) (LPred sym ast)))))
+proveGoalsOffline _adapter _opts _ctx Nothing = return Nothing
+proveGoalsOffline adapter opts ctx (Just gs0) = do
+  goalNum <- newIORef (0, 0, 0)
+  unless hasUnsatCores $ do
```
Comment (robdockins) [inline crux/src/Crux/Goal.hs]: Probably don't need this warning for offline solving, which doesn't appear to support unsat cores at all.

File: crux/src/Crux/Goal.hs [hunk 3]
```diff
@@ -129,24 +135,121 @@ proveToGoal _ allAsmps p pr =
...
+          let sym = ctx ^. ctxSymInterface
+          assumptions <- WI.andAllOf sym (L.folded . id) (fmap (^. labeledPred) (mconcat assumptionsInScope))
+          goal <- notPred sym (p ^. labeledPred)
+          -- NOTE: We don't currently provide a method for capturing the output
+          -- sent to offline solvers.  We would probably want a file per goal.
+          let logData = WS.defaultLogData
+          mres <- withTimeout $ WS.solver_adapter_check_sat adapter sym logData [assumptions, goal] $ \satRes ->
+            case satRes of
+              Unsat _ -> do
+                atomicModifyIORef' goalNum (\(x, u, s) -> ((x, u, s+1), ()))
```
Comment (robdockins) [inline crux/src/Crux/Goal.hs]: I think you're incrementing the wrong counter here, and below

File: crux/src/Crux/Goal.hs (hunk 2 above)
Comment (travitch) [reply crux/src/Crux/Goal.hs]: Thanks

File: crux/src/Crux/Goal.hs (hunk 3 above)
Comment (travitch) [reply crux/src/Crux/Goal.hs]: Whoops, thanks.  I changed it to use a record with named fields to make this mistake harder to repeat

**Comment by travitch** [comment]
That would be pretty easy, but seems like it would probably be a breaking API change - where do you think is the right place to slot that in?

**Comment by robdockins** [comment]
Well, I admit, I haven't thought through the details.  But I'd think just return the `ProcessedGoals` record along with the `Goals` structure.

File: what4/src/What4/Utils/Process.hs [hunk 4]
```diff
@@ -75,6 +75,6 @@ withProcessHandles path args mcwd action = do
               (\(_ :: SomeException) -> return ())
   let onError (_,_,_,ph) = do
         -- Interrupt process; suppress any exceptions that occur.
-        catch (interruptProcessGroupOf ph) (\(_ :: SomeException) -> return ())
+        catch (terminateProcess ph) (\(_ :: SomeException) -> return ())
```
Comment (atomb) [inline what4/src/What4/Utils/Process.hs]: What's the rationale for this change? The documentation suggests that it might be an improvement, but potentially not on Windows. Have you done any experiments there?

File: crux/src/Crux.hs [hunk 5]
```diff
@@ -233,83 +290,188 @@ data ProgramCompleteness
...
+        case mgt of
+          Nothing -> return ()
+          Just gt -> modifyIORef glsRef (Seq.|> gt)
+  when (simVerbose cruxOpts > 1) $ do
+    say "Crux" "Simulation complete."
+  when (isProfiling cruxOpts) $ writeProf profInfo
+  (,) <$> readIORef compRef <*> readIORef glsRef
+
+isProfiling :: CruxOptions -> Bool
+isProfiling cruxOpts = profileCrucibleFunctions cruxOpts || profileSolver cruxOpts
```
Comment (atomb) [inline crux/src/Crux.hs]: All of the simulator refactoring above is great, and just makes things cleaner overall.

File: crux/src/Crux.hs [hunk 6]
```diff
@@ -117,53 +130,97 @@ showVersion l = outputLn ("crux version: " ++ Crux.version ++ ", " ++
...
+    unsatCores | yicesMCSat cruxOpts = NoUnsatFeatures
+               | otherwise           = ProduceUnsatCores
+
+    withOnlineBackendFM floatRepr =
+      case selectedSolver of
+        CCS.Yices -> withYicesOnlineBackend floatRepr nonceGen unsatCores $ \sym -> do
+          symCfg sym yicesEnableMCSat (yicesMCSat cruxOpts)
+          case goalTimeout cruxOpts of
+            Just s -> symCfg sym yicesGoalTimeout (floor s)
+            Nothing -> return ()
```
Comment (atomb) [inline crux/src/Crux.hs]: I wish we had a way to reduce some of the duplication related to setting timeouts here. It makes me wonder whether a more global setting, like `solverInteractionFile`, would be appropriate. It could have a fixed set of units (seconds?) and solvers would internally have to calculate their own timeouts based on that.

File: what4/src/What4/Utils/Process.hs (hunk 4 above)
Comment (travitch) [reply what4/src/What4/Utils/Process.hs]: This is catching a missed case from #400, which made this change more widely.  Briefly, `interruptProcessGroupOf` is basically Ctrl+C, while `terminateProcess` is `kill -15` (soft terminate, not `kill -9`). I made the change because some solvers have special (and incorrect) handling of Ctrl+C that makes it not terminate the process.  The one that bit me was that yices tries to be fancy with Ctrl+C handling, but actually doesn't support Ctrl+C interrupts in MC-SAT mode.

File: crux/src/Crux.hs (hunk 6 above)
Comment (travitch) [reply crux/src/Crux.hs]: Oh my yes, that would be nice.
